dodiscover/constraint/fcialg.py

```diff
@@ -553,10 +553,19 @@ def _apply_rule8(self, graph: EquivalenceClass, u: Column, a: Column, c: Column)
         if graph.has_edge(c, a, graph.circle_edge_name) and graph.has_edge(
             a, c, graph.directed_edge_name
         ):
-            # check that A -> u or A -o u
-            condition_one = graph.has_edge(a, u, graph.directed_edge_name) or graph.has_edge(
+            # check that A -> u
+            condition_one_Adirectu = graph.has_edge(
+                a, u, graph.directed_edge_name
+            ) and not graph.has_edge(u, a, graph.circle_edge_name)
+            # check that A -o u
```

Comment:
```suggestion
            # check that A -o u
            # Note: this is not possible without first running R5-7 because a tail with a circle edge
            # would not occur through any of the other rules.
```